lib/client/src/jup.rs

```diff
@@ -0,0 +1,325 @@
+        let bids_data: &Vec<u8> = accounts_map.get(&self.market.bids).unwrap();
+        self.bids = BookSide::try_deserialize(&mut bids_data.as_slice()).unwrap();
+
+        let asks_data = accounts_map.get(&self.market.asks).unwrap();
+        self.asks = BookSide::try_deserialize(&mut asks_data.as_slice()).unwrap();
+
+        let event_queue_data = accounts_map.get(&self.market.event_queue).unwrap();
+        self.event_queue = EventQueue::try_deserialize(&mut event_queue_data.as_slice()).unwrap();
+
+        let oracle_data = accounts_map.get(&self.market.oracle).unwrap();
+        let oracle_acc = &AccountOracle {
+            data: oracle_data,
+            key: self.market.oracle,
+        };
+
+        let timestamp: u64 = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Failed to get system time")
+            .as_secs();
+        self.oracle_price = self.market.oracle_price(oracle_acc, timestamp)?;
+
+        Ok(())
+    }
+
+    fn quote(&self, quote_params: &QuoteParams) -> Result<Quote> {
+        let side = if quote_params.input_mint == self.market.quote_mint {
+            Side::Bid
+        } else {
+            Side::Ask
+        };
+
+        let order = &Order {
+            side,
+            max_base_lots: quote_params.max_base_lots,
+            max_quote_lots_including_fees: quote_params.max_quote_lots_including_fees,
+            client_order_id: 0,
+            time_in_force: 0,
+            self_trade_behavior: DecrementTake,
+            params: OrderParams::Market,
+        };
+
+        let bids_ref = RefCell::new(self.bids);
+        let asks_ref = RefCell::new(self.asks);
+        let book = Orderbook {
+            bids: bids_ref.borrow_mut(),
+            asks: asks_ref.borrow_mut(),
+        };
+
+        let timestamp: u64 = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Failed to get system time")
+            .as_secs();
+
+        let order_amounts: Amounts =
+            iterate_book(book, order, &self.market, self.oracle_price, timestamp)?;
+        let (in_amount, out_amount) = if side == Side::Bid {
+            (
+                order_amounts.total_quote_taken_native,
+                order_amounts.total_base_taken_native,
+            )
+        } else {
+            (
+                order_amounts.total_base_taken_native,
+                order_amounts.total_quote_taken_native,
+            )
+        };
+
+        Ok(Quote {
+            in_amount,
+            out_amount,
+            fee_mint: self.market.quote_mint,
+            fee_amount: order_amounts.fee,
+            price_impact_pct: order_amounts.price_impact.into(),
+            ..Quote::default()
+        })
+    }
+}
+
+pub struct Amounts {
+    pub total_base_taken_native: u64,
+    pub total_quote_taken_native: u64,
+    pub fee: u64,
+    pub price_impact: i64,
+}
+
+fn iterate_book(
+    book: Orderbook,
+    order: &Order,
```

Comment:
most of the order fields are useless, it'd probably be clearer to pass only the ones being used